news/2020-09-15-CDI_Lite_extension.adoc

```diff
@@ -0,0 +1,287 @@
+
+One issue we raised in the last post is that this build-time approach is not compatible with contemporary CDI portable extension model. Therefore, with existing CDI features we can only imagine CDI Lite without extension mechanism. Knowing the great popularity of portable extensions and their importance for integrators, we decided to revisit what could be done in that direction.
+
+We have worked to create a new extension SPI that could be used in Lite and today we are sharing this proposal with you. It is still a work in progress but should be a good enough starting point to design an extension mechanism for CDI lite.
+
+When we started thinking about a lite API design space, we mostly constrained ourselves to one particularly common use case: annotation transformations. Many things that Portable Extensions allow can be achieved simply by adding an annotation here or removing an annotation there. We know that other transformations are necessary, but the API for annotation transformation is the most developed proposal we have. Other bits and pieces are in various stages of construction, especially registering synthetic components. So, please, bear with us and submit feedback! (*TODO* where?)
+
+Before we start, we’d also like to note that in this text, we assume the reader (that is, you!) is familiar with CDI and preferably also Portable Extensions, as we expect some knowledge and don’t explain everything. We also make references to Portable Extensions on several occasions.
+
+== Extensions
+
+In our proposal, Build Compatible Extensions are simply methods annotated with an _extension annotation_. Extension annotations correspond to phases in which extensions are processed. There are 4 phases:
+
+. *@Discovery*: adding classes to application, registering custom contexts
+. *@Enhancement*: transforming annotations
+. *@Synthesis*: registering synthetic beans and observers
+. *@Validation*: performing custom validation
+
+There are some constraints we put on these methods (such as: they must be `public`), but they should be pretty obvious and shouldn’t be limiting anyone. The container automatically finds extensions and invokes them when time is right.
+When exactly are the extensions invoked can’t be defined in too much detail, because we want implementations to be able to invoke them at build time (e.g. during application compilation) or runtime (e.g. during application deployment). It certainly happens before the application starts. Extensions in earlier phases are guaranteed to run before extensions in later phases.
+
+
+Extension methods can declare an arbitrary number of parameters, all of which are supplied by the container. There’s a set of predefined types of parameters for each phase, and all implementations would have to support that. We’re also thinking of designing an SPI that would allow anyone to contribute support for other types of parameters.
```

Comment:
It seems you have copied an older version of my article, because when I worked on finalizing it, I made extra care to avoid the term "extension method". It's a commonly used name for a particular language feature (present in C# or Kotlin, but not in Java), so I think we should avoid it.

Reply:
We used an up to date one, it's just that I edited it thinking that "extension method" in this context would be clearer. Guess I was wrong :D

Reply:
Ah gotcha :-)